### ISSUE TITLE
DataTablesJS Component

### ISSUE THREAD

**Issue by hasansezertasan** [pydantic/FastUI#80]
Hi,

I discovered Starlette Admin in these last months and started to use in production. It uses [DataTables/DataTables](https://github.com/DataTables/DataTables) for tables which is quite powerfull.

My proposal is basically this: a component/field to use DataTables within FastUI.

**Comment by samuelcolvin** [comment]
Maybe, but what specifically do you need from DataTablesJS?

**Comment by hasansezertasan** [comment]
Here are some examples: [DataTables example - Zero configuration](https://datatables.net/examples/basic_init/zero_configuration.html), [DataTables example - Ajax sourced data](https://datatables.net/examples/data_sources/ajax.html)

It basically creates a table with pagination, search and sorting features. It has a lot of customization options and its free. The best part is that we can easily seed the table with JSON data. We can set pagination endpoints in it and ta da ✨

Many of my applications are data-first applications with many tables that presented to the users. Plain `table` approach is not enough for good user experience which DataTablesJS brings a it.

Long story short: It's usefull and does a lot of magic based on definitions.

**Comment by hasansezertasan** [comment]
Many of my applications are data-first applications with many tables that presented to the users. Plain `table` approach is not enough for good user experience but DataTablesJS is.

Here are some examples: [DataTables example - Zero configuration](https://datatables.net/examples/basic_init/zero_configuration.html), [DataTables example - Ajax sourced data](https://datatables.net/examples/data_sources/ajax.html)

It basically creates a table with pagination, search and sorting features. It has a lot of customization options and its free. The best part is that we can easily seed the table with JSON data. We can set pagination endpoints in it and ta da ✨

Long story short: It's usefull and does a lot of magic based on definitions. I believe I "specifically need" another (a better) option  to plain tables.

**Comment by paddymul** [comment]
FWIW as a datatable developer I have a bunch of opinions, and so do DataTables.

To the point where I think FastUI including a built in datatable component is way out of scope for the aims of this project.  I think a simple HTML/bootstrap based table could work as a builtin for FastUI.  Just take a list of strings for headers, and either a 2d array for the table, or maybe a list of dicts.  

If I were recommending a generic datatable, I would look at IPyAggrid, which closely mimics the API of Ag-Grid in a jupyter widget.  My own project, buckaroo is more opinionated, built around enabling summary stats and EDA for dataframes.

Once FastUI figures out extensibility I will make a component for it based on buckaroo.  Building one for jquery datatables, or whatever other library would be nice additions.

There are a lot of projects in this space.  I made a whole explainer page about related projects in the buckaroo docs.  https://buckaroo-data.readthedocs.io/en/latest/articles/related_projects.html

**Comment by paddymul** [comment]
@samuelcolvin  Maybe some guidance for users about what should and should not be in the core FastUI package, along with your reasoning would be helpful.

**Comment by hasansezertasan** [comment]
> To the point where I think FastUI including a built in datatable component is way out of scope for the aims of this project.

You sound right but my situation here is different. I'm not a react person. I write some Vanilla JS. Let's say that I really want to use DataTablesJS, I tried to implement it in FastUI just looking at the example here: [DataTables example - Ajax sourced data](https://datatables.net/examples/data_sources/ajax.html). I can't figure out how to make it work.

That's why I opened this issue because in the end we were gonna talk about this: Using other great JavaScript Libraries with FastUI. I can think of some other libraries:

- [sparksuite/simplemde-markdown-editor](https://github.com/sparksuite/simplemde-markdown-editor)
- [TinyMCE](https://www.tiny.cloud/)
- [ckeditor/ckeditor5](https://github.com/ckeditor/ckeditor5)
- [josdejong/jsoneditor](https://github.com/josdejong/jsoneditor)
- ...

> @samuelcolvin Maybe some guidance for users about what should and should not be in the core FastUI package, along with your reasoning would be helpful.

I believe a contribution guide could be awesome.

**Comment by paddymul** [comment]
Custom components which will be necessary for third party components are discussed here:
https://github.com/pydantic/FastUI/issues/19

When I was trying to learn FastUI and get a feel for it, I created a PR that integrated my own component into core FastUI, expecting the PR would never be merged.  It still gave me a feel for many of the pieces and code that will be needed for a custom component.  You could take a look here https://github.com/pydantic/FastUI/pull/73

**Comment by hasansezertasan** [comment]
@paddymul Thank you for your answer. #88 now adds Custom Components support but I'm not sure how we will be able to start a component that initialized via JavaScript snippet.
